// @stoplight/prism-http/dist/mocker/negotiator/InternalHelpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.contentHasExamples = exports.createResponseFromDefault = exports.findResponseByStatusCode = exports.findFirstResponse = exports.findLowest2xx = exports.findDefaultContentType = exports.findBestHttpContentByMediaType = exports.findExampleByKey = exports.findFirstExample = void 0;
const accepts = require("accepts");
const contentType = require("content-type");
const O = require("fp-ts/Option");
const A = require("fp-ts/Array");
const lodash_1 = require("lodash");
const NEA = require("fp-ts/NonEmptyArray");
const Ord_1 = require("fp-ts/Ord");
const pipeable_1 = require("fp-ts/pipeable");
function findFirstExample(httpContent) {
    return pipeable_1.pipe(O.fromNullable(httpContent.examples), O.chain(NEA.fromArray), O.chain(A.head));
}
exports.findFirstExample = findFirstExample;
function findExampleByKey(httpContent, exampleKey) {
    return pipeable_1.pipe(O.fromNullable(httpContent.examples), O.chain(A.findFirst(({ key }) => key === exampleKey)));
}
exports.findExampleByKey = findExampleByKey;
function findBestHttpContentByMediaType(contents, mediaTypes) {
    const bestType = accepts({ headers: { accept: mediaTypes.join(',') } }).type(contents.map(c => c.mediaType));
    return pipeable_1.pipe(bestType, O.fromPredicate((bestType) => !!bestType), O.chain(bestType => A.findFirst(content => content.mediaType === bestType)(contents)), O.alt(() => pipeable_1.pipe(mediaTypes
        .map(mt => contentType.parse(mt))
        .filter(({ parameters }) => Object.keys(parameters).some(k => k !== 'q'))
        .map(({ type, parameters }) => ({ type, parameters: lodash_1.pick(parameters, 'q') }))
        .map(mt => contentType.format(mt)), NEA.fromArray, O.chain(mediaTypesWithNoParameters => findBestHttpContentByMediaType(contents, mediaTypesWithNoParameters)))));
}
exports.findBestHttpContentByMediaType = findBestHttpContentByMediaType;
function findDefaultContentType(contents) {
    return pipeable_1.pipe(contents, A.findFirst(content => content.mediaType === '*/*'));
}
exports.findDefaultContentType = findDefaultContentType;
const byResponseCode = Ord_1.ord.contramap(Ord_1.ordNumber, response => parseInt(response.code));
function findLowest2xx(httpResponses) {
    const first2xxResponse = pipeable_1.pipe(httpResponses, A.filter(response => /2\d\d/.test(response.code)), A.sort(byResponseCode), A.head);
    return pipeable_1.pipe(first2xxResponse, O.alt(() => createResponseFromDefault(httpResponses, 200)));
}
exports.findLowest2xx = findLowest2xx;
function findFirstResponse(httpResponses) {
    return pipeable_1.pipe(httpResponses, A.head);
}
exports.findFirstResponse = findFirstResponse;
function findResponseByStatusCode(responses, statusCode) {
    return pipeable_1.pipe(responses, A.findFirst(response => response.code.toLowerCase() === String(statusCode)));
}
exports.findResponseByStatusCode = findResponseByStatusCode;
function createResponseFromDefault(responses, statusCode) {
    return pipeable_1.pipe(responses, A.findFirst(response => response.code === 'default'), O.map(response => Object.assign({}, response, { code: statusCode })));
}
exports.createResponseFromDefault = createResponseFromDefault;
function contentHasExamples(content) {
    return !!content.examples && content.examples.length !== 0;
}
exports.contentHasExamples = contentHasExamples;
